### ISSUE TITLE
BIM+TechDraw: Problem displaying doors on plans, facades and sections.

### ISSUE THREAD

**Issue by kaiwas** [FreeCAD/FreeCAD#16450]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

We have many ways to get a house layout, but there is no convenient way to get the type of doors that are used in construction drawings.
I understand that standards may differ in different countries, but very often doors should not have overlapping lines (examples of display from standards and other construction programs).

Revit
![Scr_24-09-11-05-23-34](https://github.com/user-attachments/assets/ad3e89b2-13ca-4f9b-9fb2-db6a1f0561c0)

ArchiCAD
![Scr_24-09-11-05-24-28](https://github.com/user-attachments/assets/ca95cc4c-2aeb-4909-8d8c-028b18e6b688)

Standards in Russia (I am sure that in many countries it is similar)
![Scr_24-09-11-05-26-06](https://github.com/user-attachments/assets/e2414143-df11-4c77-beac-fc7e63d47575)

As you can see in the pictures, most often the doorway is not covered by lines.

The situation in the FreeCAD looks like this.

![Door_problem](https://github.com/user-attachments/assets/9fb5ddf9-f42b-4e47-8a36-8ee55185f557)

As you can see - the door is not open, lines are displayed in the opening.
And the options themselves look different depending on the building.

Extra lines in the opening can be removed by opening the door, for example, 100%. But in this case, an unpleasant feature arises - the door on the facade (and sections) also opens. And this is what is most often not required.

As a result, in order to have the desired type of door in the plan, it is necessary to separately generate a plan with open doors. And separately generate facades with closed doors. And this is very unproductive.

The file for tests is attached. Remove the zip extension.
[Planes_Facades.FCStd.zip](https://github.com/user-attachments/files/16965157/Planes_Facades.FCStd.zip)

Discussion on the forum.
https://forum.freecad.org/viewtopic.php?t=90409

### Full version info

```shell
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 1.0.0RC1.38642 (Git) AppImage
Build type: Release
Branch: (HEAD detached at 1.0rc1)
Hash: 60a251354e203bdb924e2190fb9f5e18b48d7362
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: Russian/Russia (ru_RU)
Installed mods: 
  * OpenTheme 2024.9.1
  * PitchedRoof
```

### Subproject(s) affected?

BIM

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by 4a3** [comment]
In order for this to be possible, it is necessary to introduce the concept of projection as a 2d document.
Then it will be possible to customize the image by replacing the element with a symbol

**Comment by 4a3** [comment]
Then the issue with the display settings for doors, windows, plumbing, cars, trees and other things can be solved at a qualitatively new level

**Comment by paullee0** [comment]
Thanks.  Anyone has more information with IFC, would like to see how this may be implemented with IFC-compatible in mind :)

**Comment by marcuspollio** [comment]
Hi there ! =D

A few comments to have a broader picture of the current state and of potential paths forwards :

You are probably aware that, until the big Arch/BIM/Native-IFC merge, there was only one common Tool for both Windows and Doors. Some of this legacy is still present today (same Window icons in the Tree view, same Window elements in the Task panel, same Window section in the Data Property view, etc.), tough the IFC Type and IFC Attributes are different now.
The considerations here are to be extended to both Tools and possibly others, as mentioned above by others.

It can be asserted that @yorikvanhavre is well aware of this situation, as solving most of the current limitations is planned in the [roadmap](https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/main/doc/roadmap2024.md). More details about the specific Window Tool redesign [can be found](https://yorik.uncreated.net/blog/2023-025-new-window-tool) as well. That said, carrying such enormous tasks cannot be a solo effort. Moreover, described as now, this issue cannot be resolved in one go, as it involves too many different bits. It needs to be divided in smaller goals/tasks.

To help define them and see how it could work in a "environment" close to FreeCAD, I invite you to test how doors and windows work in [Bonsai/BlenderBIM](https://bonsaibim.org/) (get it directly from within Blender Extensions in Preferences). There, Types and Representations based on IFC are supported and allow greater flexibility, editing and customization that what's currently possible in FreeCAD (although please note that documentation is patchy for now). As some of it is also based on IfcOpenShell, there may be interesting synergies to develop, also to cover greater use cases and local specialties and requirements.

**Comment by paullee0** [comment]
> To help define them and see how it could work in a "environment" close to FreeCAD, I invite you to test how doors and windows work in [Bonsai/BlenderBIM](https://bonsaibim.org/) (get it directly from within Blender Extensions in Preferences). There, Types and Representations based on IFC are supported and allow greater flexibility, editing and customization that what's currently possible in FreeCAD (although please note that documentation is patchy for now). As some of it is also based on IfcOpenShell, there may be interesting synergies to develop, also to cover greater use cases and local specialties and requirements.

Thanks.  Could you elaborate what can be done for FreeCAD making reference from BBIM ?

**Comment by yorikvanhavre** [comment]
Basically if I resume what you're asking:

- The ability to not display the door geometry (only the arc). That's indeed intersting, I hadn't thought about it. Shouldn't be too hard, I'll add that to my plan
- Different styles of opening representation: The current system is made so one can actually add more styles, when I do the big refactor I'll try to make that better
- Door/window tags. That's indeed needed. Wall tags are also needed. I would think, instead of embedding that in windows, it could be interesting to use the Draft label object for that, and all other kinds of "texts that are or not in a frame, that can be attached to objects, and take their value from it".

If any of you guys want to help with that, I'd be grateful if you could study what can already be achieved with Draft Labels, and what would be missing

**Comment by paullee0** [comment]
I have a few ideas on Arch Window / Door, might be different from Yorik's Roadmap.

But if everyone agree with the feature to assign 2d representation, I guess this could be done not too hard.  In fact, the Window/Door (3d) shape should better not be cut to make the 2d view everytime TechDraw needs, kind of wasting resources.

**Comment by semhustej** [comment]
> I have a few ideas on Arch Window / Door, might be different from Yorik's Roadmap.
> 
> But if everyone agree with the feature to assign 2d representation, I guess this could be done not too hard. In fact, the Window/Door (3d) shape should better not be cut to make the 2d view everytime TechDraw needs, kind of wasting resources.

Yes, in Revit it is much less resource demanding when you assign only 2D geometry for the plan view. 

I think there should be an option for user - either cut 3D model of the door if you do not want to bother with 2D representation or create 2D representation and disable the 3D model in the plan view - using contexts.

**Comment by paullee0** [comment]
So we add a new attribute e.g. PlanAnnotationPlanView which point to e.g. a Sketch (an arc with a straight, rectangle representing the door leaf)

**Comment by semhustej** [comment]
Yes, that is the simplified way.

IfcGeometricRepresentationContext is a bit more complex, as you can see from this schema:
![IFC-context](https://github.com/user-attachments/assets/8e4ef09a-a82d-42f6-9521-935852312a58)

https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcGeometricRepresentationContext.htm

I am not sure, how much FC BIM development wants to follow IFC schema or how much it wants to go it's own way. 
Following IFC schema has advantage of not inventing something, that already exists. It is also following what is most closely related to "industry standard", even though no industry standard really exist.
On the other hand, there are a lot of things in the IFC schema, that might be more complicated and labor intensive to implement in FC than what could be done without adhering to it.

But that is not a discussion for an issue report thread :)

**Comment by paullee0** [comment]
> But that is not a discussion for an issue report thread :)

Indeed.  Native-IFC seems supported by IfcOpenShell and should be same as BBIM somehow.

So, here we discuss and agree on specifically how the shape on plan, elevation and section should be handled first :)

> So we add a new attribute e.g. PlanAnnotationPlanView which point to e.g. a Sketch (an arc with a straight, rectangle representing the door leaf)

Now, more idea or agreement to use the attribute name 'PlanAnnotationPlanView' ?

Next is, should there be a 'library' of this plan/elevation/section shape, how this could be stored, and to be added by users?

**Comment by semhustej** [comment]
Yes the name seems to make sense:

<ContextType><ContextIdentifier><TargetView>

<Plan><Annotation><PLAN_VIEW>

**Comment by semhustej** [comment]
Yes the name makes sense

'ContextType' -> 'ContextIdentifier' -> 'TargetView'

'Plan' -> 'Annotation' -> 'PLAN_VIEW'

Not sure if it's better to stick with IFC's PLAN_VIEW or PlanView.

**Comment by kaiwas** [comment]
> Next is, should there be a 'library' of this plan/elevation/section shape/annotation, how this could be stored, and to be added by users?

It would be great to have different options for presentation on the plan. But for this we will have to study the most commonly used in different countries. However, they are most likely very similar.

**Comment by marcuspollio** [comment]
Hi there !

Thank you all for the inspiring discussion ! and @semhustej for providing more details about Bonsai/BlenderBIM and IfcOpenShell, it's more less what I had in mind.

For the Representation side of it, we could indeed use and assign to Objects either "automatic" cuts or projections onto planes of the said object, or custom Sketches or Draft Wires, based on the Context logic discussed by you above.

To "store" that, using [Types](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcTypeObject.htm) (or its FreeCAD equivalent for non-IFC fallback) could be interesting, so to reuse them in the same project or in other projects as well.

As for sharing these library assets, there may be some interesting synergies with a [FreeCAD-wide system](https://github.com/FreeCAD/FreeCAD-developer-meetings/blob/main/Minutes/minutes-2024-09-14.md) to be developed (read the third point). We could imagine using local Types or getting remote ones shared by us all in a community effort, to cover greater use cases and regional specifications.

For the Tag/Label aspect, yes, a generic labeling mechanism that can retrieve any referenced info (of any data type, string, float, int, etc.) from linked Objects would be great. It should be implemented to fully take advantage of [Property Sets](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcPropertySet.htm), so this generic label stuff would maybe come after PSets works I suppose. But we should open another issue specifically for that. 

Finally, on the View and TechDraw side, ideally, the whole TD system should be overhauled so as to give a proper meaning to the generated representations and its element, from the linked Objects, because at present it works in a rather "blind" way. But this is a huge task that needs to be discussed in another issue too.

**Comment by paullee0** [comment]
Thanks, good to have more and more concrete ideas :)  Anything needs discussion back at the forum?
 
> For the Representation side of it, we could indeed use and assign to Objects either "automatic" cuts or projections onto normal planes of the said object, or custom Sketches or Draft Wires, based on the Context logic discussed by you above.

Needs a few new properties ?

1. SymbolGeneration - [ 'Cut Door', 'PlanAnnotationPlan_View]
2. 'PlanAnnotationPlan_View' (Object?) - PropertyLink to a 'Library / Type' object, or a Sketch or Draft Object ?
(Or the shape of the object user directed be simply copied and saved in a Part::PropertyPartShape / Part::PropertyTopoShapeList )

> To "store" that, using [Types](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcTypeObject.htm) (or its FreeCAD equivalent for non-IFC fallback) could be interesting, so to reuse them in the same project or in other projects as well.

I remember there is a Part_Library but I do not use it.  This is shomehow too big and slow to download etc.  Is it 'compatible' to IFC usage here or

**Comment by marcuspollio** [comment]
Hi @paullee0 

To be honest, I am really not familiar with the code base, so I cannot comment regarding the implementation per se. In any case, I guess a certain balance has to be found between keeping as much backward compatibility as possible and developing generic new methods that work well FreeCAD-wide (e.g. in a dependency/linking logic).

Do you mean the [FreeCAD library](https://github.com/FreeCAD/FreeCAD-library) ? In a Native-IFC context, using only simple text-based IFC (in one or several of its [data file formats](https://en.wikipedia.org/wiki/Industry_Foundation_Classes#File_Formats)) instead of a compound file format like FCStd may be easier. But there should be a mechanism so all features work whether or not the IFC approach is used.

Another related aspect that needs consideration, but slightly off-topic (sorry =P) is the scale and Level of Detail. We currently have the "[Hi Res](https://wiki.freecad.org/Arch_Component)" Property in the Component base of most BIM Objects. While convenient, it is quite makeshift and not well integrated as a proper design/workflow feature. What I mean is dealing with scale iterations and LoD changes is an everyday routine. Although that's for another issue, improving what we are discussing here (Representations, Types, etc.) should also facilitate future scale/LoD management improvements.

**Comment by semhustej** [comment]
> To "store" that, using [Types](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcTypeObject.htm) (or its FreeCAD equivalent for non-IFC fallback) could be interesting, so to reuse them in the same project or in other projects as well.

The idea of "Types" of doors, windows and other objects, which could be than easily used across different projects, is sound. This is the typical approach used in other BIM software.


> Another related aspect that needs consideration, but slightly off-topic (sorry =P) is the scale and Level of Detail.

I did not mention LOD (Level of Detail/Development), because that brings another layer of complexity as you mentioned :). But it is good to think about it during the door/window system refactor, so that it would be possible to implement LOD in the future.

**Comment by yorikvanhavre** [comment]
> But if everyone agree with the feature to assign 2d representation, I guess this could be done not too hard. In fact, the Window/Door (3d) shape should better not be cut to make the 2d view everytime TechDraw needs, kind of wasting resources.

I'm honestly not convinced of separating 2D and 3D outputs. It seems to me a lot of code complication, and therefore less control and more bugs. I am totally willing to be proved wrong on this of course :)

If we want to go that way (BIM objects can output separate 3D and 2D shapes), this should be made a separate, generic mechanism, because it could be also useful for ex. for pieces of furniture

Regarding BBIM and how it works in IFC, indeed that's an interesting thing that a same object can have different representations  in different contexts, ex: one for 3D and one for 2D. But then, instead of separating into two "rendering pipelines", why not cache a 2D "section view"?

**Comment by semhustej** [comment]
@paullee0 I am no expert on IFC, so there may be mistakes in my description, but from what I understand (I will be talking about Windows):

If an object has **IfcWindowType** class assigned, it is saved as somewhat of a "definition block" hidden in the IFC project. Instances/occurrences (individual IfcWindows) of that type are placed in the project. When the "definition block" is changed, all instances change accordingly. In FC terms, it would be something like a clone or link.
IfcWindowType can have Properties (which is something like an attribute or variable), which can be either Inherited Type property or an Occurrence property. Inherited type properties have values for all in instances of the window the same, and are defined in the IfcWindowType. Occurrence properties can have values different for each instance=occurrence.

So you could have for instance IfcWindowType 'Sigle pane wooden frame' with type property "Window frame material"=wood and let's say occurrence property "Smart window sensor", which could be set to yes or no for every IfcWindow in the project independently.

I am not sure, whether windows in IFC schema can be parametric, so that width/height could be set as occurrence parameter different for each occurence/instance in the project or whether size of the window has to be type parameter.

I am not sure, whether the 2D plan representation (PlanAnnotationPlan_View) is part of the IfcWindowType. I would guess that these PlanAnnotationPlan_View contexts do not exist on their own - so one can not be assigned to multiple different window types.

This is how it looks like in bonsaiBIM project tree:
![BBIM-instances](https://github.com/user-attachments/assets/c329dc34-70a5-42c8-ac4d-abb050f29f8a)

If there are mistakes in my description, please don't hesitate to correct me :)

**Comment by paullee0** [comment]
Thanks, I have just checked IfcArchitect's video e.g.  https://www.youtube.com/watch?v=2Q_wtBKa8Dc, seems one difference is user, after creation of a door / window, can re-select the 'Type'.

In FreeCAD BIM, user have about 10 window/door preset 'type', user have all the flexibility to tweak the underlying Sketch (e.g. make it arch-shape, trapezoid, oval etc.)  and sub-component ( e.g. any nos. of frames, louvers etc.), but not a simple switch to select other preset 'types'.  Not sure what other things 'Type' would allow (e.g. in BBIM).

This should becomes a separate thread :)

In the meantime, seems nothing about 'Type' prevent the implementation of the separate 2D representation here, agreed?

**Comment by semhustej** [comment]
> This should becomes a separate thread :)

Yes, we have come quite off-topic :)

> In the meantime, seems nothing about 'Type' prevent the implementation of the separate 2D representation here, agreed?
I do not think type prevents having 2D representation as a separate entity.

**Comment by yorikvanhavre** [comment]
Types is another matter completely indeed.

Types is something simple, it's like a "template" object, for ex. A WindowType is exactly the same as a Window, but it's not inserted in the model. If a window uses it as its type, all the properties that are not overridden by the window are taken from the type.

In my BIM roadmap I plan to introduce types everywhere. I have a bit of code ready already, need to clean that up first.
